book/src/pil/types.md

```diff
@@ -0,0 +1,245 @@
+documentation purposes, since they cannot occur in identities.
+
+They allow the following operators:
+
+- `+`: string concatenation
+
+### Tuple
+
+Type name: `(..., ..., ...)`
+
+Tuples are complex types that are composed from other types, either zero or two or more. There is no tuple type with a single element (`(int)` is the same as `int`).
+The empty tuple type is written as `()`.
+
+Examples include `(int, int)` (a pair of integers) and `((fe[], int), ())`
+(a tuple consisting of a tuple that contains an array of field elements and an integer
+and an empty tuple).
+
+Tuples values are constructed using parentheses: `(1, 2)` constructs the tuple that consists of
+a one and a two.
+
+Tuples do not allow any operators.
+
+### Array
+
+Type name: `_[]`
+
+Arrays are statically or dynamically-sized collections of elements each of the same type
+denoted for example as `int[]` (a dynamically-sized array of integers) or `int[2]` (an array of integers with static size two).
+Array values can be constructed inline using `[1, 2]` (the array containing the
+two elements one and two).
+
+The built-in function `std::array::len` can be used to retrieve the length of an array (statically or dynamically sized)
+and the elements of an array `a` can be accessed using `a[0]`, `a[1]`, etc.
+
+Arrays allow the following operators:
+
+- `+`: array concatenation
+- `_[]`: array index access, the index needs to be a non-negative integer that is less than the length of the array, otherwise a runtime error is triggered
+
+
+### Function
+
+Type name: `T1, T2, ..., Tn -> T0`
+
+Function type names are for example denoted as `int, fe -> int` or `-> int`.
+Note that `(int, fe) -> int` is a function that takes a single tuple as parameter
+while `int, fe -> int` takes two parameters of type integer and field element.
+
+Functions can be constructed using the lambda expression notation.
+For example `|x, y| x + y` returns a function that performs addition.
+The lambda expression `|| 7` is a function that returns a constant (has no parameters).
+Lambda functions can capture anything in their environment and thus form closures.
+
+Due to the dynamic typing of powdr-pil, it is often not possible to exactly determine the type of a lambda expression.
+
+Functions allow the following operators:
+
+- `_(...)`: function evaluation
+
+Powdr-pil is usually side-effect free, but there are some built-in functions that have
+side-effects:
+These are `std::debug::print` and `std::check::panic` and all functions that call them.
+Expressions are eagerly evaluated from left to right.
+
+### Expression
+
+Type name: `expr`
+
+Expressions are the elements of the algebraic expressions used in identities.
+Expressions cannot be explicitly constructed. At statement level / in identity contexts,
+references to columns are implicitly converted to expressions, as are integers and field elements.
+
+Example:
+```rust
+let x;
+let y;
+let f = || x + y;
+let g = || 7;
+f() == g();
```

Comment:
Two equal signs?

Reply:
yea I've brought this up before but I'm still not convinced also

Reply:
https://github.com/powdr-labs/powdr/issues/958